src/aspire/basis/basis.py

```diff
@@ -74,24 +74,55 @@ def evaluate(self, v):
         Evaluate coefficient vector in basis
 
         :param v: A coefficient vector (or an array of coefficient vectors)
-            to be evaluated. The first dimension must equal `self.count`.
+            to be evaluated. The first dimension must correspond to the number of
+            coefficient vectors, while the second must correspond to `self.count`
         :return: The evaluation of the coefficient vector(s) `v` for this basis.
-            This is an array whose first dimensions equal `self.z` and the
-            remaining dimensions correspond to dimensions two and higher of `v`.
+            This is an Image or a Volume object containing one image/volume for each
+            coefficient vector, and of size `self.sz`.
         """
+        if v.dtype != self.dtype:
+            logger.warning(
+                f"{self.__class__.__name__}::evaluate"
+                f" Inconsistent dtypes v: {v.dtype} self: {self.dtype}"
+            )
+
+        if self.ndim == 2:
+            return Image(self._evaluate(v))
+        elif self.ndim == 3:
+            return Volume(self._evaluate(v))
+
+    def _evaluate(self, v):
         raise NotImplementedError("subclasses must implement this")
 
     def evaluate_t(self, v):
         """
         Evaluate coefficient in dual basis
 
-        :param v: The coefficient array to be evaluated. The first dimensions
-            must equal `self.sz`.
-        :return: The evaluation of the coefficient array `v` in the dual
+        :param v: An Image or Volume object whose size matches `self.sz`.
+        :return: The evaluation of the Image or Volume object `v` in the dual
             basis of `basis`.
-            This is an array of vectors whose first dimension equals `self.count`
-            and whose remaining dimensions correspond to higher dimensions of `v`.
+            This is an array of vectors whose first dimension equals the number of
+            images/volumes in `v`. and whose second dimension is `self.count`.
         """
+        if v.dtype != self.dtype:
+            logger.warning(
+                f"{self.__class__.__name__}::evaluate_t"
+                f" Inconsistent dtypes v: {v.dtype} self: {self.dtype}"
+            )
+
+        if not isinstance(v, Image) and not isinstance(v, Volume):
+            if self.ndim == 2:
+                _class = Image
+            elif self.ndim == 3:
+                _class = Volume
+            logger.warning(
+                f"{self.__class__.__name__}::evaluate_t"
+                f" passed numpy array instead of {_class}."
+            )
+            v = _class(v)
```

Comment:
They way I read the intent of the PR was that the inner methods (`_evaluate`) and (`_evaluate_t`) should only deal with ndarrays. Why are we casting here?

Reply:
The `evaluate_t` methods for the various bases are a bit inconsistent as far as whether they expect `Image/Volume` or arrays. The choice was many modifications to all the specific functions versus just a few and passing an `Image/Volume`. I didn't initially intend to dig super deep into each individual method with this PR, just impose the overall structure and enforce the return / input types. 

Do you think it would be better to make all those changes in this PR?

Reply:
Ok let's make an issue.